File: ch01/data1.5.jl
include("../utils.jl")
using Random, DataFrames,Pipe,CSV,PrettyTables,GLMakie,TableTransforms
Random.seed!(343434)

desc=Lock5Table(56,"LifeExpectancyVehicles","Cofounding-Variables",["Year","LifeExpectancy","Vehicles"])

df=@pipe CSV.File("./data/$(desc.name).csv")|>DataFrame

@pt first(df,20)

#= 
┌───────┬────────────────┬──────────┐
│  Year │ LifeExpectancy │ Vehicles │
│ Int64 │        Float64 │  Float64 │
├───────┼────────────────┼──────────┤
│  1970 │           70.8 │    108.4 │
│  1971 │           71.1 │    113.0 │
│  1972 │           71.2 │    118.8 │
│  1973 │           71.4 │    125.7 │
│  1974 │           72.0 │    129.9 │
│  1975 │           72.6 │    132.9 │
│  1976 │           72.9 │    138.5 │
│  1977 │           73.3 │    142.1 │
│  1978 │           73.5 │    148.4 │
│  1979 │           73.9 │    151.9 │
│  1980 │           73.7 │    155.8 │
│  1981 │           74.1 │    158.3 │
│  1982 │           74.5 │    159.6 │
│  1983 │           74.6 │    163.7 │
│  1984 │           74.7 │    166.2 │
│  1985 │           74.7 │    171.7 │
│   ⋮   │       ⋮        │    ⋮     │
└───────┴────────────────┴──────────┘
=#
table=eachcol(df)
data=table|>Scale(2,3, low=0, high=1)

function  plot_res()
    fig=Figure()
    ax=Axis(fig[1,1])
    Box(fig[1,1];color = (:orange,0.1),strokewidth=0.5)
    lines!(ax,data.Year,data.LifeExpectancy;label="$(desc.feature[2])",linewidth=3)
    lines!(ax,data.Year,data.Vehicles;label="$(desc.feature[3])",linewidth=3)
    axislegend(ax)
    fig
end

#fig=plot_res();
#save("./ch01/data1.5-1.png",fig)

function  plot_res2()
    fig=Figure()
    ax=Axis(fig[1,1];xlabel = "LifeExpectancy", ylabel = "Vehicles")
    Box(fig[1,1];color = (:orange,0.1),strokewidth=0.5)
    scatter!(ax,df.LifeExpectancy,df.Vehicles;markersize=12,color=(:lightgreen,0.2),strokecolor = :black, strokewidth =2)
    
    fig
end

fig=plot_res2()
save("./ch01/imgs/data1.5-2.png",fig)

File: marimo/ch01/data1.5.py
import marimo

__generated_with = "0.24.0"
app = marimo.App(width="medium")


@app.cell
def _():
    import marimo as mo
    import sys
    from pathlib import Path
    import numpy as np
    import polars as pl
    import matplotlib.pyplot as plt
    from scipy import stats
    from great_tables import GT
    _p = Path(mo.notebook_location()).resolve().parent
    while _p != _p.parent and not (_p / "lock5.py").exists():
        _p = _p.parent
    if str(_p) not in sys.path:
        sys.path.insert(0, str(_p))
    import lock5

    return GT, lock5, mo, pl, plt


@app.cell
def _(mo):
    mo.md(r"""
    # Data 1.5 — Life expectancy and vehicles (confounding variables)

    Port of `UnitA/ch01/data1.5.jl`. Read `LifeExpectancyVehicles`, min-max scale
    `LifeExpectancy` and `Vehicles` to `[0, 1]` (`TableTransforms.Scale(2, 3, low=0, high=1)`),
    then plot both series against `Year` and a scatter of the two variables.
    """)
    return


@app.cell
def _(GT, lock5, mo):
    desc = lock5.Lock5Table(56, "LifeExpectancyVehicles", "Cofounding-Variables",
                            ["Year", "LifeExpectancy", "Vehicles"])
    df = lock5.load_csv(desc.name)
    mo.Html(GT(df.head(20)).tab_header(title=desc.name).as_raw_html())
    return desc, df


@app.cell
def _(df, pl):
    # min-max scale columns 2 and 3 (LifeExpectancy, Vehicles) to [0, 1]
    scaled = df.select([
        pl.col("Year"),
        ((pl.col("LifeExpectancy") - pl.col("LifeExpectancy").min())
         / (pl.col("LifeExpectancy").max() - pl.col("LifeExpectancy").min())).alias("LifeExpectancy"),
        ((pl.col("Vehicles") - pl.col("Vehicles").min())
         / (pl.col("Vehicles").max() - pl.col("Vehicles").min())).alias("Vehicles"),
    ])
    return (scaled,)


@app.cell
def _(desc, plt, scaled):
    def plot_line():
        fig, ax = plt.subplots(figsize=(7, 4))
        ax.plot(scaled["Year"], scaled["LifeExpectancy"], label=desc.feature[1], linewidth=3)
        ax.plot(scaled["Year"], scaled["Vehicles"], label=desc.feature[2], linewidth=3)
        ax.set_xlabel("Year")
        ax.set_ylabel("scaled value")
        ax.set_title("Life expectancy and vehicles over time (scaled)")
        ax.legend()
        return fig

    plot_line()
    return


@app.cell
def _(df, plt):



    fig2, ax2 = plt.subplots(figsize=(7, 4))
    ax2.scatter(df["LifeExpectancy"], df["Vehicles"], s=50, color="lightgreen",
               alpha=0.5, edgecolors="black", linewidths=1)
    ax2.set_xlabel("LifeExpectancy")
    ax2.set_ylabel("Vehicles")
    ax2.set_title("Vehicles vs Life expectancy")
    fig2
    return


if __name__ == "__main__":
    app.run()
